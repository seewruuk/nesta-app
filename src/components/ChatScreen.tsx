import React, { useState, useEffect } from 'react';
import { View, Text, FlatList, TextInput, TouchableOpacity } from 'react-native';
import { useRouter } from 'expo-router';
import { useStateContext } from '@/src/contexts/StateContext';
import { Message as MsgType } from '@/src/types/Message';

export default function ChatScreen() {
    const router = useRouter();
    const {
        state: { currentUserId, messages },
        addMessage,
        deleteMessage,
        markAsRead,
    } = useStateContext();

    const [text, setText] = useState('');

    const chatMsgs = [...messages].sort(
        (a, b) => new Date(a.date).getTime() - new Date(b.date).getTime()
    );

    useEffect(() => {
        if (!currentUserId) {
            router.replace('/login');
        }
    }, [currentUserId]);

    useEffect(() => {
        if (!currentUserId) return;
        chatMsgs.forEach(msg => {
            if (msg.receiverId === currentUserId && !msg.isRead) {
                markAsRead(msg.id);
            }
        });
    }, [chatMsgs, currentUserId]);

    if (!currentUserId) return null;

    const handleSend = () => {
        if (!text.trim()) return;
        const newMsg: MsgType = {
            id: `msg${Date.now()}`,
            senderId: currentUserId,
            receiverId: '', // unused in static view
            message: text.trim(),
            date: new Date().toISOString(),
            isRead: false,
            status: 'sent',
        };
        addMessage(newMsg);
        setText('');

        setTimeout(() => {
            const reply: MsgType = {
                id: `bot${Date.now()}`,
                senderId: 'bot',
                receiverId: '',
                message: 'tak',
                date: new Date().toISOString(),
                isRead: true,
                status: 'read',
            };
            addMessage(reply);
        }, 5000);
    };

    const renderItem = ({ item }: { item: MsgType }) => {
        const isMine = item.senderId === currentUserId;
        const isBot = item.senderId === 'bot';
        return (
            <View
                style={{
                    alignSelf: isMine ? 'flex-end' : 'flex-start',
                    backgroundColor: isMine
                        ? '#DCF8C5'
                        : isBot
                            ? '#FFE0B2'
                            : '#FFF',
                    margin: 4,
                    padding: 8,
                    borderRadius: 8,
                    borderWidth: isBot ? 1 : 0,
                    borderColor: isBot ? '#FF9800' : 'transparent',
                }}
            >
                <Text style={isBot ? { fontStyle: 'italic' } : {}}>{item.message}</Text>
                <Text style={{ fontSize: 10, textAlign: 'right' }}>
                    {new Date(item.date).toLocaleTimeString()}
                </Text>
                <TouchableOpacity onPress={() => deleteMessage(item.id)}>
                    <Text style={{ fontSize: 10, color: 'red' }}>Usuń</Text>
                </TouchableOpacity>
            </View>
        );
    };

    return (
        <View style={{ flex: 1, padding: 8, paddingBottom: 150, paddingTop: 50 }}>
            <FlatList
                data={chatMsgs}
                keyExtractor={item => item.id}
                renderItem={renderItem}
            />
            <View style={{ flexDirection: 'row', alignItems: 'center' }}>
                <TextInput
                    style={{ flex: 1, borderWidth: 1, borderRadius: 4, padding: 8 }}
                    placeholder="Napisz wiadomość..."
                    value={text}
                    onChangeText={setText}
                />
                <TouchableOpacity onPress={handleSend} style={{ marginLeft: 8 }}>
                    <Text>Wyślij</Text>
                </TouchableOpacity>
            </View>
        </View>
    );
}
