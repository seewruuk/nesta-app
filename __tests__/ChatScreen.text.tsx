import React from 'react';
import { render, fireEvent, act } from '@testing-library/react-native';
import ChatScreen from '@/src/components/ChatScreen';

jest.mock('expo-router', () => ({
    useRouter: () => ({ replace: jest.fn() }),
}));

const mockAddMessage = jest.fn();
const mockDeleteMessage = jest.fn();
const mockMarkAsRead = jest.fn();
const mockUpdateStatus = jest.fn();

jest.mock('@/src/contexts/StateContext', () => ({
    useStateContext: () => ({
        state: { currentUserId: 'user1', messages: [] },
        addMessage: mockAddMessage,
        deleteMessage: mockDeleteMessage,
        markAsRead: mockMarkAsRead,
        updateMessageStatus: mockUpdateStatus,
    }),
}));

describe('<ChatScreen />', () => {
    beforeEach(() => {
        jest.useFakeTimers();
        mockAddMessage.mockClear();
    });

    it('match the snapshot if there is no message', () => {
        const tree = render(<ChatScreen />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('cakk addMessage after tapping send message and show auto reply after 5s timeout', () => {
        const { getByPlaceholderText, getByText } = render(<ChatScreen />);

        const input = getByPlaceholderText('Napisz wiadomość...');
        const sendBtn = getByText('Wyślij');

        fireEvent.changeText(input, 'Test wiadomość');
        fireEvent.press(sendBtn);

        expect(mockAddMessage).toHaveBeenCalledTimes(1);
        const sentPayload = mockAddMessage.mock.calls[0][0];
        expect(sentPayload.message).toBe('Test wiadomość');
        expect(sentPayload.senderId).toBe('user1');
        expect(sentPayload.status).toBe('sent');

        act(() => {
            jest.advanceTimersByTime(5000);
        });
        expect(mockAddMessage).toHaveBeenCalledTimes(2);
        const botPayload = mockAddMessage.mock.calls[1][0];
        expect(botPayload.message).toBe('tak');
        expect(botPayload.senderId).toBe('bot');
        expect(botPayload.status).toBe('read');
    });
});
