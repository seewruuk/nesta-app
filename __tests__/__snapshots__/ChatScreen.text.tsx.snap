// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChatScreen /> match the snapshot if there is no message 1`] = `
<View
  style={
    {
      "flex": 1,
      "padding": 8,
    }
  }
>
  <RCTScrollView
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View />
  </RCTScrollView>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="Napisz wiadomość..."
      style={
        {
          "borderRadius": 4,
          "borderWidth": 1,
          "flex": 1,
          "padding": 8,
        }
      }
      value=""
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "marginLeft": 8,
          "opacity": 1,
        }
      }
    >
      <Text>
        Wyślij
      </Text>
    </View>
  </View>
</View>
`;
